src/ucs/memory/numa.c

```diff
@@ -72,3 +94,263 @@ int ucs_numa_node_of_cpu(int cpu)
 }
 
 #endif
+
+static inline uint32_t ucs_numa_distance_hash_key(const ucs_numa_node_t node1,
+                                                  const ucs_numa_node_t node2)
+{
+    static const uint8_t shift = sizeof(ucs_numa_node_t) * 8;
+
+    return (ucs_max(node1, node2) << shift) | ucs_min(node1, node2);
+}
+
+/**
+ * Iterate the directory entries in @param path
+ * of the form @param prefix<index> and return the
+ * maximum index.
+ * <index> is a positive number.
+ *
+ * @param [in]  path          Directory path
+ * @param [in]  prefix        Consider only entries starting with this prefix
+ * @param [in]  prefix_length Prefix length
+ *
+ * @return Maximum index of the entries starting with prefix or -1 if error.
+ */
+static int ucs_numa_get_max_dirent(const char *path, const char *prefix,
+                                   const size_t prefix_length)
+{
+    int max_index = 0;
+    DIR *dir;
+    struct dirent *de;
+    int idx;
+
+    dir = opendir(path);
+    if (!dir) {
+        return -1;
+    } else {
+        while ((de = readdir(dir)) != NULL) {
+            if (strncmp(de->d_name, prefix, prefix_length)) {
+                continue;
+            }
+
+            idx = strtoul(de->d_name + prefix_length, 0, 0);
+            if (max_index < idx) {
+                max_index = idx;
+            }
+        }
+
+        closedir(dir);
+    }
+
+    return max_index;
+}
+
+int ucs_numa_num_configured_nodes()
+{
+    static int max_node = -1;
+
+    if (max_node == -1) {
+        max_node = ucs_numa_get_max_dirent(UCS_NUMA_NODES_DIR_PATH, "node", 4);
+        if (max_node == -1) {
+            ucs_error("Failed to parse sysfs nodes dir");
+            max_node = UCS_NUMA_DEFAULT_NODE;
+        } else if (max_node > UCS_NUMA_MAX_NODE) {
+            ucs_error("max NUMA node is out of range");
+            max_node = UCS_NUMA_DEFAULT_NODE;
+        }
+    }
+
+    return max_node;
+}
+
+int ucs_numa_num_configured_cpus()
+{
+    static int max_cpu = -1;
+
+    ucs_spin_lock(&ucs_numa_global_ctx.lock);
+    if (max_cpu == -1) {
+        max_cpu = ucs_numa_get_max_dirent(UCS_NUMA_CORES_DIR_PATH, "cpu", 3);
```

Comment:
The call to `get_max_dirent` and subsequent logic repeats 3 times.
WDYT of exporting this to a macro:
```
        max_cpu = ucs_numa_get_max_dirent(UCS_NUMA_CORES_DIR_PATH, "cpu", 3);
        if (max_cpu == -1) {
            ucs_error("Failed to parse sysfs CPU's dir");
            max_cpu = 0;
        } else if (max_cpu > __CPU_SETSIZE - 1) {
            ucs_error("CPU set size is out of range");
            max_cpu = 0;
        }
```

Then will have instead something like:
```
if (max_cpu == -1) {
   UCS_NUMA_GET_MAX_DIRENT(max_cpu, UCS_NUMA_CORES_DIR_PATH,
                           "cpu", 3, __CPU_SETSIZE - 1, 0);
   max_cpu += 1;
}
```
and similarly in the 2 other places.

Reply:
We can do that.
Maybe the macro can also use sizeof instead of passing the prefix length? WDYT?